tests/unit/fem/test_estimator.cpp

```diff
@@ -52,6 +50,216 @@ double NonsmoothSolutionZ(const mfem::Vector& x)
+      GridFunction u_gf(&fespace);
+      u_gf.ProjectCoefficient(u_analytic);
+
+      DiffusionIntegrator di;
+      NewZienkiewiczZhuEstimator estimator(di, u_gf);
+
+      auto &local_errors = estimator.GetLocalErrors();
+      for (int i=0; i<local_errors.Size(); i++)
+      {
+         REQUIRE(local_errors(i) < 1e-10);
+      }
+      REQUIRE(estimator.GetTotalError() < 1e-10);;
+   }
+
+   SECTION("Perfect Approximation Y")
+   {
+      FunctionCoefficient u_analytic(testhelper::SmoothSolutionY);
+      GridFunction u_gf(&fespace);
+      u_gf.ProjectCoefficient(u_analytic);
+
+      DiffusionIntegrator di;
+      NewZienkiewiczZhuEstimator estimator(di, u_gf);
+
+      auto &local_errors = estimator.GetLocalErrors();
+      for (int i=0; i<local_errors.Size(); i++)
+      {
+         REQUIRE(local_errors(i) < 1e-10);
+      }
+      REQUIRE(estimator.GetTotalError() < 1e-10);
+   }
+
+   SECTION("Perfect Approximation Z")
+   {
+      FunctionCoefficient u_analytic(testhelper::SmoothSolutionZ);
+      GridFunction u_gf(&fespace);
+      u_gf.ProjectCoefficient(u_analytic);
+
+      DiffusionIntegrator di;
+      NewZienkiewiczZhuEstimator estimator(di, u_gf);
+
+      auto &local_errors = estimator.GetLocalErrors();
+      for (int i=0; i<local_errors.Size(); i++)
+      {
+         REQUIRE(local_errors(i) < 1e-10);
+      }
+      REQUIRE(estimator.GetTotalError() < 1e-10);
+   }
+
+   SECTION("Nonsmooth Approximation X")
+   {
+      FunctionCoefficient u_analytic(testhelper::NonsmoothSolutionX);
+      GridFunction u_gf(&fespace);
+      u_gf.ProjectCoefficient(u_analytic);
+
+      DiffusionIntegrator di;
+      NewZienkiewiczZhuEstimator estimator(di, u_gf);
+
+      auto &local_errors = estimator.GetLocalErrors();
+      for (int i=0; i<local_errors.Size(); i++)
+      {
+         REQUIRE(local_errors(i) >= 0.0);
+      }
+      REQUIRE(estimator.GetTotalError() > 0.0);
+   }
+
+   SECTION("Nonsmooth Approximation Y")
+   {
+      FunctionCoefficient u_analytic(testhelper::NonsmoothSolutionY);
+      GridFunction u_gf(&fespace);
+      u_gf.ProjectCoefficient(u_analytic);
+
+      DiffusionIntegrator di;
+      NewZienkiewiczZhuEstimator estimator(di, u_gf);
+
+      auto &local_errors = estimator.GetLocalErrors();
+      for (int i=0; i<local_errors.Size(); i++)
+      {
+         REQUIRE(local_errors(i) >= 0.0);
+      }
+      REQUIRE(estimator.GetTotalError() > 0.0);
+   }
+
+   SECTION("Nonsmooth Approximation Z")
+   {
+      FunctionCoefficient u_analytic(testhelper::NonsmoothSolutionZ);
+      GridFunction u_gf(&fespace);
+      u_gf.ProjectCoefficient(u_analytic);
+
+      DiffusionIntegrator di;
+      NewZienkiewiczZhuEstimator estimator(di, u_gf);
+
+      auto &local_errors = estimator.GetLocalErrors();
+      for (int i=0; i<local_errors.Size(); i++)
+      {
+         REQUIRE(local_errors(i) >= 0.0);
+      }
+      REQUIRE(estimator.GetTotalError() > 0.0);
```

Comment:
These tests seem to basically assert that the error for smooth solutions should be zero, and for non-smooth solutions nonzero. Wouldn't the original ZZ estimator pass these tests? Is there a stronger test that the new estimator would pass but the original ZZ estimator would not?

Reply:
I think it would be ideal to set up a minimal test showing correct convergence, which fails for the original ZZ. Hopefully "minimal" means maybe orders {1,2,3} on a very small initial mesh, with just a couple of refinements?